File: src/org/foonugget/kdict/ui/KoreanDictionaryActivity.java

package org.foonugget.kdict.ui;

import java.util.ArrayList;
import java.util.List;

import org.foonugget.kdict.R;
import org.foonugget.kdict.R.id;
import org.foonugget.kdict.data.Dictionary;
import org.foonugget.kdict.data.SearchHistory;
import org.foonugget.kdict.data.WordMatch;

import roboguice.activity.RoboActivity;
import roboguice.inject.InjectView;
import android.os.Bundle;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import com.google.inject.Inject;

public class KoreanDictionaryActivity extends RoboActivity {

    @SuppressWarnings("unused")
    private final static String TAG = KoreanDictionaryActivity.class
            .getSimpleName();

    enum SearchType {
        Exact,
        Contains,
        Each
    }

    @InjectView(id.searchText)
    private EditText mSearchTextField;

    /* Contains search history or search results depending on the context */
    @InjectView(id.wordList)
    private ListView mWordListView;

    @Inject
    private Dictionary mDict;

    @Inject
    private SearchHistory mSearchHistory;

    @Inject
    private WordSearchListAdapter.Factory wordSearchListAdapterFactory;

    @Inject
    private SearchHistoryListAdapter.Factory searchHistoryListAdapterFactory;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        populateSearchHistory();
    }

    private final OnClickListener onClickSearchHistoryEntry = new OnClickListener() {
        @Override
        public void onClick(View v) {
            String searchText = (String) ((TextView) v).getText();
            doSearch(SearchType.Exact, searchText);
            mSearchTextField.setText(searchText);
        }
    };

    private void populateSearchHistory() {
        List<String> searchList = mSearchHistory.getSearchHistoryAll();
        SearchHistoryListAdapter adapter = searchHistoryListAdapterFactory.create(searchList,
                onClickSearchHistoryEntry);
        mWordListView.setAdapter(adapter);
    }

    public void onClickContains(View view) {
        doSearch(SearchType.Contains);
    }

    public void onClickEach(View view) {
        doSearch(SearchType.Each);
    }

    public void onClickExact(View view) {
        doSearch(SearchType.Exact);
    }

    public void doSearch(SearchType type) {
        String searchText = mSearchTextField.getText().toString();
        doSearch(type, searchText);
    }

    private void doSearch(SearchType type, String searchText) {
        List<WordMatch> matchesList;
        switch (type) {
            case Contains:
                matchesList = mDict.wordSearchContains(searchText);
                break;
            case Each:
                matchesList = mDict.wordSearchEach(searchText);
                break;
            case Exact:
                matchesList = mDict.wordSearch(searchText);
                break;
            default:
                matchesList = new ArrayList<WordMatch>();
        }

        WordSearchListAdapter adapter = wordSearchListAdapterFactory.create(matchesList);
        mWordListView.setAdapter(adapter);
        mSearchHistory.addSearchWordToHistory(searchText);
    }

}

File: app/src/main/java/org/foonugget/kdict/ui/KoreanDictionaryActivity.kt
package org.foonugget.kdict.ui

import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import org.foonugget.kdict.R
import org.foonugget.kdict.data.Dictionary
import org.foonugget.kdict.data.SearchHistory
import org.foonugget.kdict.data.WordMatch
import org.koin.android.ext.android.inject

class KoreanDictionaryActivity : AppCompatActivity() {

    private val mDict: Dictionary by inject()
    private val mSearchHistory: SearchHistory by inject()
    private val wordSearchListAdapterFactory: WordSearchListAdapter.Factory by inject()
    private val searchHistoryListAdapterFactory: SearchHistoryListAdapter.Factory by inject()

    private lateinit var mSearchTextField: EditText
    private lateinit var mWordListView: ListView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        mSearchTextField = findViewById(R.id.searchText)
        mWordListView = findViewById(R.id.wordList)
        populateSearchHistory()
    }

    private val onClickSearchHistoryEntry = View.OnClickListener { v ->
        val searchText = (v as TextView).text.toString()
        doSearch(SearchType.Exact, searchText)
        mSearchTextField.setText(searchText)
    }

    private fun populateSearchHistory() {
        val searchList = mSearchHistory.getSearchHistoryAll()
        val adapter = searchHistoryListAdapterFactory.create(searchList, onClickSearchHistoryEntry)
        mWordListView.adapter = adapter
    }

    @Suppress("UNUSED_PARAMETER")
    fun onClickContains(view: View) {
        doSearch(SearchType.Contains)
    }

    @Suppress("UNUSED_PARAMETER")
    fun onClickEach(view: View) {
        doSearch(SearchType.Each)
    }

    @Suppress("UNUSED_PARAMETER")
    fun onClickExact(view: View) {
        doSearch(SearchType.Exact)
    }

    fun doSearch(type: SearchType) {
        val searchText = mSearchTextField.text.toString()
        doSearch(type, searchText)
    }

    private fun doSearch(type: SearchType, searchText: String) {
        if (searchText.isBlank()) return

        val matchesList: List<WordMatch> = when (type) {
            SearchType.Contains -> mDict.wordSearchContains(searchText)
            SearchType.Each -> mDict.wordSearchEach(searchText)
            SearchType.Exact -> mDict.wordSearch(searchText)
        }

        val adapter = wordSearchListAdapterFactory.create(matchesList)
        mWordListView.adapter = adapter
        mSearchHistory.addSearchWordToHistory(searchText.trim())
    }

    enum class SearchType {
        Exact,
        Contains,
        Each
    }
}
